### PR TITLE
Impact Particle Effects

### PR DESCRIPTION
**Pull request by tyeagg**
## About The Pull Request
bullet impacts now use particles

https://user-images.githubusercontent.com/98577688/233021143-7745377b-2dbc-4a1f-b587-81a4cbc1d4cf.mp4
## Why It's Good For The Game
looks cool also may fix the whole bullet ping getting stuck
## Changelog
:cl:
add: particle effects for bullet impacts
imageadd: debris for bullet impacts
code: variables to determine what debris an object spews out when its impacted by a projectile
/:cl:

### CODE REVIEW

**Comment by Al-1ce** [comment]
pretty !

**Comment by XElectricX** [comment]
Holy hell bless with another Tyeagg particle PR. Speed merge. Probably much better for performance too and stops the weird bug of impact effects not qdel-ing themselves.

File: code/game/atoms.dm [hunk 1]
```diff
@@ -115,6 +115,14 @@
...
 	///The color this atom will be if we choose to draw it on the minimap
 	var/minimap_color = MINIMAP_SOLID
+	///Icon state of debris when impacted by a projectile
+	var/debris = FALSE
+	///Amount of debris particles
+	var/debris_amount = 8
+	///Velocity of debris particles
+	var/debris_velocity = -15
+	///If the debris particles have a bloom filter
+	var/debris_bloom = FALSE
```
Comment (TiviPlus) [inline code/game/atoms.dm:125]: Gross here is
A) Atom vars, could this be an element or something instead
B) FALSE/string mixing
C) string should be defines

File: code/modules/projectiles/projectile.dm [hunk 2]
```diff
@@ -1297,31 +1298,73 @@ So if we are on the 32th absolute pixel coordinate we are on tile 1, but if we a
...
+		if(P.ammo.flags_ammo_behavior & AMMO_SOUND_PITCH)
+			pitch = 55000
+		playsound(src, P.ammo.sound_bounce, 50, 1, frequency = pitch)
+	addtimer(CALLBACK(src, PROC_REF(remove_ping), src, smoke_visuals, debris_visuals), 0.7 SECONDS)
+
+/atom/proc/remove_ping(atom/hit, smoke_visuals, debris_visuals)
+	if(!QDELETED(hit))
+		QDEL_NULL(smoke_visuals)
+		if(debris_visuals)
+			QDEL_NULL(debris_visuals)
```
Comment (TiviPlus) [inline code/modules/projectiles/projectile.dm:1367]: hit == src?
also wont timers not run on qdeleted objects anyway

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 3]
```diff
@@ -21,50 +21,100 @@
...
+
 //unsorted miscellaneous temporary visuals
 /obj/effect/temp_visual/dir_setting/bloodsplatter
 	icon = 'icons/effects/blood.dmi'
 	duration = 0.5 SECONDS
 	randomdir = FALSE
-	layer = BELOW_MOB_LAYER
+	layer = ABOVE_MOB_LAYER
 	var/splatter_type = "splatter"
+	alpha = 175
```
Comment (TiviPlus) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:53]: new vars above new definitions

**Review by TiviPlus — CHANGES_REQUESTED** [review]
rest ltgm

File: code/datums/elements/debris.dm [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+/datum/element/debris/Detach(datum/source, force)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ATOM_BULLET_ACT)
+
+/datum/element/debris/proc/register_for_impact(datum/source, obj/projectile/proj)
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, PROC_REF(on_impact), source, proj)
+
+/datum/element/debris/proc/on_impact(datum/source, obj/projectile/P)
```
Comment (TiviPlus) [inline code/datums/elements/debris.dm:60]: Autodocs

**Review by TiviPlus — APPROVED** [review]

**Comment by Lumipharon** [comment]
Particle effects are definitely the future, but during the current TM these are way over the top to the point of being distracting. A group of marines shooting in the direction of a wall creates a giant continuous fountain of debris exploding out.

**Comment by FosterGrime-Heymynameslouie** [comment]
> Particle effects are definitely the future, but during the current TM these are way over the top to the point of being distracting. A group of marines shooting in the direction of a wall creates a giant continuous fountain of debris exploding out.

sounds pretty cool to look at honestly

**Comment by pilotmans** [comment]
This is beautiful, this is just quality.

**Pull request merged**